source/types.ts

```diff
@@ -0,0 +1,226 @@
+
+	**Note:** The values in `defaults` will overwrite the `default` key in the `schema` option.
+	*/
+	defaults?: Readonly<T>;
+
+	/**
+	[JSON Schema](https://json-schema.org) to validate your config data.
+
+	Under the hood, the JSON Schema validator [ajv](https://github.com/epoberezkin/ajv) is used to validate your config. We use [JSON Schema draft-07](http://json-schema.org/latest/json-schema-validation.html) and support all [validation keywords](https://github.com/epoberezkin/ajv/blob/master/KEYWORDS.md) and [formats](https://github.com/epoberezkin/ajv#formats).
+
+	You should define your schema as an object where each key is the name of your data's property and each value is a JSON schema used to validate that property. See more [here](https://json-schema.org/understanding-json-schema/reference/object.html#properties).
+
+	@example
+	```
+	import Conf = require('conf');
+
+	const schema = {
+		foo: {
+			type: 'number',
+			maximum: 100,
+			minimum: 1,
+			default: 50
+		},
+		bar: {
+			type: 'string',
+			format: 'url'
+		}
+	};
+
+	const config = new Conf({schema});
+
+	console.log(config.get('foo'));
+	//=> 50
+
+	config.set('foo', '1');
+	// [Error: Config schema violation: `foo` should be number]
+	```
+
+	**Note:** The `default` value will be overwritten by the `defaults` option if set.
+	*/
+	schema?: { [TProp in keyof T]: Schema; };
+
+	/**
+	Name of the config file (without extension).
+
+	Useful if you need multiple config files for your app or module. For example, different config files between two major versions.
+
+	@default 'config'
+	*/
+	configName?: string;
+
+	/**
+	You only need to specify this if you don't have a package.json file in your project or if it doesn't have a name defined within it.
+
+	Default: The name field in the `package.json` closest to where `conf` is imported.
+	*/
+	projectName?: string;
+
+	/**
+	You only need to specify this if you don't have a package.json file in your project or if it doesn't have a version defined within it.
+
+	Default: The name field in the `package.json` closest to where `conf` is imported.
+	*/
+	projectVersion?: string;
+
+	/*
+	_Don't use this feature until [this issue](https://github.com/sindresorhus/conf/issues/92) has been fixed._
+
+	You can use migrations to perform operations to the store whenever a version is changed.
+
+	The `migrations` object should consist of a key-value pair of `'version': handler`. The `version` can also be a [semver range](https://github.com/npm/node-semver#ranges).
+
+	@example
+	```
+	import Conf = require('conf');
+
+	const store = new Conf({
+		migrations: {
+			'0.0.1': store => {
+				store.set('debugPhase', true);
+			},
+			'1.0.0': store => {
+				store.delete('debugPhase');
+				store.set('phase', '1.0.0');
+			},
+			'1.0.2': store => {
+				store.set('phase', '1.0.2');
+			},
+			'>=2.0.0': store => {
+				store.set('phase', '>=2.0.0');
+			}
+		}
+	});
+	```
+	*/
+	migrations?: Migrations<T>;
```

Comment:
All of these should be prefixed with `readonly`.

Reply:
I'll fix this with `Readonly<>` otherwise I won't be able to set them internally in the constructor.